### DISCUSSION TITLE
eglot-workspace-configuration / initializationOptions setup

### DISCUSSION THREAD

**Discussion by rsoeldner** [joaotavora/eglot#1148] [General]
I'm exploring emacs and eglot in general. Using the `.dir-locals.el` worked pretty well:
```
((nil
  . ((eglot-workspace-configuration
      . (:haskell (:plugin
		   (:stan
		    (:globalOn :json-false))))))))
```

I tried using the `initializationOptions` as follows:
```
(defun haskellInit (&optional srv)
  (:haskell (:plugin (:stan (:globalOn :json-false))))  )

(use-package eglot
  :ensure t
  :hook (haskell-mode . eglot-ensure)
  :config
  (add-to-list 'eglot-server-programs '(haskell-mode . ("haskell-language-server-wrapper" "--lsp" :initializationOptions (haskellInit))))
  )
```

resulting in the following (within the `Message` Window):
```
Loading /usr/share/emacs/site-lisp/site-start.d/asy-init.el (source)...done
Loading /usr/share/emacs/site-lisp/site-start.d/desktop-entry-mode-init.el (source)...done
For information about GNU Emacs and the GNU system, type C-h C-a.
Error in post-command-hook ((closure ((--cl-maybe-connect-- closure #1 nil (progn (remove-hook 'post-command-hook --cl-maybe-connect-- nil) (let ((G4 buffer)) (if (buffer-live-p G4) (save-current-buffer (set-buffer G4) (if eglot--managed-mode nil (apply #'eglot--connect (eglot--guess-contact)))))))) (buffer . #<buffer Check.hs>) eglot--managed-mode eglot-lsp-context markdown-fontify-code-blocks-natively t) nil (progn (remove-hook 'post-command-hook --cl-maybe-connect-- nil) (let ((G4 buffer)) (if (buffer-live-p G4) (save-current-buffer (set-buffer G4) (if eglot--managed-mode nil (apply #'eglot--connect (eglot--guess-contact))))))))): (wrong-type-argument consp nil)
Mark set
```

I struggle with this issue, are there existing examples somewhere?

**Comment by joaotavora** [comment]
Hi @rsoeldner , thanks for your question. 

I hope you don't mind, but I'm going to use your message as an instrument to illustrate some needed advice to new users of Emacs and Eglot. So even though I'm following your code, I'm not disparaging your honest attempts or skills, I'm talking to a more general audience of Eglot users who commonly face similar problems, very often derived from copy-pasting and adapting little snippets of code that they don't fully understand.

First don't use `eglot-ensure`!   Where do so many people get advice on doing so?  Please, only use it *after* you are 110% sure that `M-x eglot`  works correctly (`M-x` is a way of running commands within Emacs, you likely know that).  And -- if you can -- share with the bloggers or reddit enthusiasts where you grabbed this info to stop suggesting it.

Here, I think that (at least part of) the problem is that the list element after `:initializationOptions` is itself a list with only the symbol `haskellInit`.  This is probably not what you want. If you want to use auxiliary functions to tidy up your initialization code you should learn how backquoted lists work vs quoted lists.

Also, I suggest as a learning exercise to also stop using `use-package`, too. It's an unneeded tool that obscures the Lisp code that is being run, and which configuration steps are happening, how hooks are added to and what forms are being evaluated versus what forms are being simply read. In short `use-package` adds unneeded magic and makes learning Lisp foundations more difficult. And learning Lisp foundations (even if just for solving this little Haskell business) seems to be something you need to improve on.

Which reminds me that `haskellInit` is trying to call the `:haskell` function which probably also does not exist. You likely mean to quote it. Again, learn about quoted lists in Lisp. And why does `haskellInit` take an unused `srv` param? Stop cargo cults! 😄 

Again, this problem would have been easier to detect and debug if you weren't using `eglot-ensure`.  It's like getting into an F1 car before learning how to drive a mini (the mini being `M-x eglot`). And it's about as useful as a F1 car, meaning not really much, just a kind of a noisy show off.

Finally, here's a version of your code that possibly works:

```lisp
(require 'eglot)
(defun haskell-init ()   '(:haskell (:plugin (:stan (:globalOn :json-false))))  )
(add-to-list 'eglot-server-programs `(haskell-mode . ("haskell-language-server-wrapper" "--lsp" :initializationOptions ,(haskell-init))))
```

I say *possibly* because you are substituting a set of LSP "workspace configuration options" for a set of "initialization options". And these **may or may not** be interchangeable, and that depends entirely on the particular LSP server you are using.

**Reply by joaotavora** [reply]
Just as a curiosity, here's what ChatGPT has to say about this:
![image](https://user-images.githubusercontent.com/387011/211023197-68b4b6a7-857b-4078-b66a-43a76f64de72.png)

I can't really commend it for suggesting `use-package` ,but it gets high marks for suggesting `M-x eglot` and none of `eglot-ensure`.

(I do _not_ endorse ChatGPT for coming up with these snippets in general.  It was extremely lucky here and the task I gave it is reasonably simple.  It is precisely the tool that doesn't understand what it's doing, it's just linguistically recycling bits and pieces gleaned off the internet).

**Reply by rsoeldner** [reply]
@joaotavora I highly appreciate your effort and will take your advice seriously :+1:
